sam/pileup.go

```diff
@@ -0,0 +1,273 @@
+package sam
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/chromInfo"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/dna"
+	"log"
+)
+
+// Pile stores the number of each base observed across multiple reads.
+// The number of a particular base can be retrieved by using the dna.Base
+// as the index for Pile.Count. E.g. the number of reads with A is Count[dna.A].
+//
+// Linked list mechanics:
+// In the pileup function the Pile struct is part of a circular doubly-linked
+// list in which Pile.next/prev is the next/prev contiguous position. Discontiguous
+// positions are not allowed. When adding base information to the pile struct, the
+// pointer to the start of the read position should be kept, and a second pointer
+// will increment along the list as bases are added. The pointer to the beginning
+// of the list should only be incremented when a position is passed and set to zero.
+type Pile struct {
+	RefIdx   int
+	Pos      uint32
+	Count    [13]int        // Count[dna.Base] == Number of observed dna.Base
+	InsCount map[string]int // key is insertion sequence as string
+
+	touched bool // true if Count or InsCount has been modified
+	// touched is used as a quick check to see whether a pile struct
+	// has been used. This value is used to avoid unnecessary allocations
+	// in resetPile, and is used in checkSend to avoid sending zeroed structs
+	// that happen to pass the user-defined filters.
+
+	next *Pile
+	prev *Pile
+}
+
+// GoPileup inputs a channel of coordinate sorted Sam structs and generates
+// a Pile for each base. The Pile is send through the return channel when
+// the Pile position is no longer being updated.
+//
+// The input readFilter functions must all be true for a read to be included.
+// The input pileFilter functions can be likewise be used to filter the output.
+// All reads/piles are included if filters == nil.
+func GoPileup(reads <-chan Sam, header Header, includeNoData bool, readFilters []func(s Sam) bool, pileFilters []func(p Pile) bool) <-chan Pile {
+	if header.Metadata.SortOrder[0] != Coordinate {
+		log.Fatal("ERROR: (GoPileup) input sam/bam must be coordinate sorted")
+	}
+	pileChan := make(chan Pile, 1000)
```

Comment:
Correct, the buffer is arbitrarily large. Because we are using goroutines there are two threads operating concurrently: a 'reader' thread, and a 'worker' thread. Depending on the read either the reader or the worker may finish first. If the channel was unbuffered then the thread that finished first would be forced to wait until the other thread is finished before working on the next read. The buffer makes it so once one of the thread finishes, it can go on to the next read without waiting for the other thread.